server/src/main/java/org/opensearch/cluster/node/DiscoveryNodes.java

```diff
@@ -393,74 +393,77 @@ public DiscoveryNode resolveNode(String node) {
-                            }
-                        } else if (DiscoveryNodeRole.INGEST_ROLE.roleName().equals(matchAttrName)) {
-                            if (Booleans.parseBoolean(matchAttrValue, true)) {
-                                resolvedNodesIds.addAll(ingestNodes.keys());
-                            } else {
-                                resolvedNodesIds.removeAll(ingestNodes.keys());
-                            }
-                        } else if (DiscoveryNode.COORDINATING_ONLY.equals(matchAttrName)) {
-                            if (Booleans.parseBoolean(matchAttrValue, true)) {
-                                resolvedNodesIds.addAll(getCoordinatingOnlyNodes().keys());
-                            } else {
-                                resolvedNodesIds.removeAll(getCoordinatingOnlyNodes().keys());
+                    } else if (nodeExists(nodeId)) {
+                        resolvedNodesIds.add(nodeId);
+                    } else {
+                        for (DiscoveryNode node : this) {
+                            if ("_all".equals(nodeId)
+                                || Regex.simpleMatch(nodeId, node.getName())
+                                || Regex.simpleMatch(nodeId, node.getHostAddress())
+                                || Regex.simpleMatch(nodeId, node.getHostName())) {
+                                resolvedNodesIds.add(node.getId());
                             }
-                        } else {
-                            for (DiscoveryNode node : this) {
-                                for (DiscoveryNodeRole role : Sets.difference(node.getRoles(), DiscoveryNodeRole.BUILT_IN_ROLES)) {
-                                    if (role.roleName().equals(matchAttrName)) {
-                                        if (Booleans.parseBoolean(matchAttrValue, true)) {
-                                            resolvedNodesIds.add(node.getId());
-                                        } else {
-                                            resolvedNodesIds.remove(node.getId());
+                        }
+                        int index = nodeId.indexOf(':');
+                        if (index != -1) {
+                            String matchAttrName = nodeId.substring(0, index);
+                            String matchAttrValue = nodeId.substring(index + 1);
+                            if (DiscoveryNodeRole.DATA_ROLE.roleName().equals(matchAttrName)) {
+                                if (Booleans.parseBoolean(matchAttrValue, true)) {
+                                    resolvedNodesIds.addAll(dataNodes.keys());
+                                } else {
+                                    resolvedNodesIds.removeAll(dataNodes.keys());
+                                }
+                            } else if (DiscoveryNodeRole.MASTER_ROLE.roleName().equals(matchAttrName)
+                                || DiscoveryNodeRole.CLUSTER_MANAGER_ROLE.roleName().equals(matchAttrName)) {
```

Comment:
Although there are lots of lines changes in this file, most of them are indent change, this is the second actual changes (out of 2):
`|| DiscoveryNodeRole.CLUSTER_MANAGER_ROLE.roleName().equals(matchAttrName)`